expect: Improve report when mock-spy matcher fails, part 4

pedrottimark opened the pull request:
## Summary

For `toHave*ReturnedWith` matchers:

* Display matcher name in regular black instead of dim color
* Replace sentences with `Expected` and `Received` labels
* For negative matcher, repeat `not` following `Expected` label
* At end of report, display `Received number of returns`
* If any calls didn’t return, display `Received number of calls`

Display up to 3 call results as context, so testers can diagnose problems: in code, or test, or both

Your **feedback** is especially welcome about questions:

1. For special case of exactly **one call**, is it clearer to display `Received value: whatever` on one line instead of two? (see pictures in following comments)
2. Is the following **context** relevant?

Read the following descriptions with the qualifier **if they exist**:

| matcher | context |
| ---: | :--- |
| `.not.toHaveReturnedWith` | first 3 results that are equal to expected value |
| `.toHaveReturnedWith` | first 3 results, including `threw an error` |
| `.not.toHaveLastReturnedWith` | last and next-to-last results |
| `.toHaveLastReturnedWith` | last result and either nearest preceding value that is equal to expected; otherwise, next-to-last result |
| `.not.toHaveNthReturnedWith` | nth and adjacent results |
| `.toHaveNthReturnedWith` | nth and either nearest preceding/following value that is equal to expected; otherwise, adjacent results |

## Test plan

Updated 102 snapshots

| | long name | short name |
| ---: | :--- | :--- |
| 30 | `toHaveReturnedWith` | `toReturnWith` |
| 30 | `toHaveLastReturnedWith` | `lastReturnedWith` |
| 42 | `toHaveNthReturnedWith` | `nthReturnedWith` |

See also pictures in following comments

Example pictures baseline at left and **improved at right**

pedrottimark commented:
negative `toHaveReturnedWith` assertions

Here is the first example of question 1:

<img width="1000" alt="ReturnWith true 1 1" src="https://user-images.githubusercontent.com/11862657/61417696-0c9b3500-a8c6-11e9-9c98-ad330e0a34ab.png">

Display up to 3 examples of calls which returned the not-expected value:

<img width="1000" alt="ReturnWIth true 1 3 4" src="https://user-images.githubusercontent.com/11862657/61417701-102ebc00-a8c6-11e9-9837-fb9c95a98818.png">

Notice that baseline dittos the expected asymmetric matcher instead of a received value:

<img width="1000" alt="ReturnWith true 3 4 4" src="https://user-images.githubusercontent.com/11862657/61417705-13c24300-a8c6-11e9-8b94-73d21ea87762.png">

pedrottimark commented:
positive `toHaveReturnedWith` assertions

<img width="1000" alt="ReturnWith false 0 0" src="https://user-images.githubusercontent.com/11862657/61417829-92b77b80-a8c6-11e9-90e7-95087f37705d.png">

<img width="1000" alt="ReturnWith false 1 1" src="https://user-images.githubusercontent.com/11862657/61417834-9b0fb680-a8c6-11e9-8d12-67db537aaa83.png">

Display (up to) the first 3 results, including `has not returned yet`

<img width="1000" alt="ReturnWith false 2 3" src="https://user-images.githubusercontent.com/11862657/61417846-a367f180-a8c6-11e9-89d7-78ccb32a3b5e.png">

Here is another example of question 1:

<img width="1000" alt="ReturnWith false 4 4" src="https://user-images.githubusercontent.com/11862657/61417849-a7940f00-a8c6-11e9-953d-174b1d46b726.png">

pedrottimark commented:
negative `toHaveLastReturnedWith` assertions

<img width="1000" alt="LastReturn true 1 1" src="https://user-images.githubusercontent.com/11862657/61418067-4882ca00-a8c7-11e9-8bff-452ce695b43c.png">

last result and either nearest preceding value that is equal to expected; otherwise, in this picture **next-to-last result**:

<img width="1000" alt="LastReturn true 1 3" src="https://user-images.githubusercontent.com/11862657/61418075-4de01480-a8c7-11e9-8303-0ed5c7472ecb.png">

pedrottimark commented:
positive `toHaveLastReturnedWith` assertions

<img width="1000" alt="LastReturn false 0 0" src="https://user-images.githubusercontent.com/11862657/61418162-9dbedb80-a8c7-11e9-972c-c20c8889b893.png">

<img width="1000" alt="LastReturn false 0 1a" src="https://user-images.githubusercontent.com/11862657/61418166-a2838f80-a8c7-11e9-8408-9634af0a9ea3.png">

<img width="1000" alt="LastReturn false 0 1b" src="https://user-images.githubusercontent.com/11862657/61418168-a57e8000-a8c7-11e9-87ae-b5dbc7b040c5.png">

last and next-to-last results:

<img width="1000" alt="LastReturn false 0 3" src="https://user-images.githubusercontent.com/11862657/61418172-a9120700-a8c7-11e9-83e0-f7c4a9a823c5.png">

last result and nearest preceding value that is equal to expected:

<img width="1000" alt="LastReturn false 1 3" src="https://user-images.githubusercontent.com/11862657/61418173-afa07e80-a8c7-11e9-88a5-f7d1aacd79ed.png">

pedrottimark commented:
negative `toHaveNthReturnedWith` assertions

Here is the picture that caused me to ask question 1:

<img width="1000" alt="NthReturn true 1 1" src="https://user-images.githubusercontent.com/11862657/61418342-37868880-a8c8-11e9-8e83-aa2e6eef46b8.png">

nth and adjacent results, preceding but there is no following:

<img width="1000" alt="NthReturn true 2 2" src="https://user-images.githubusercontent.com/11862657/61418346-3a817900-a8c8-11e9-921b-36d51ce840d8.png">

nth and adjacent results, both preceding and following:

<img width="1000" alt="NthReturn true 10 12" src="https://user-images.githubusercontent.com/11862657/61418353-3e150000-a8c8-11e9-9fb6-ae1aa0ee2725.png">

pedrottimark commented:
positive `toHaveNthReturnedWith` assertions

<img width="1000" alt="NthReturn false 1 0" src="https://user-images.githubusercontent.com/11862657/61418488-aebc1c80-a8c8-11e9-9cbe-78bfcfd92359.png">

<img width="1000" alt="NthReturn false 1 1" src="https://user-images.githubusercontent.com/11862657/61418492-b1b70d00-a8c8-11e9-8110-9a3da8a60755.png">

How well can you pick out the `Call n` number in the lines under `Received values`?

<img width="1000" alt="NthReturn false 1 2" src="https://user-images.githubusercontent.com/11862657/61418493-b4196700-a8c8-11e9-9ffb-cce1c09f37e3.png">

nth and adjacent results:

<img width="1000" alt="NthReturn false 9 10 - 1" src="https://user-images.githubusercontent.com/11862657/61418498-b7acee00-a8c8-11e9-9650-369cc3602721.png">

nth and nearest preceding value that is equal to expected:

<img width="1000" alt="NthReturn false 9 10 - 2" src="https://user-images.githubusercontent.com/11862657/61418503-ba0f4800-a8c8-11e9-8659-88aec0f14426.png">

nth and nearest following value that is equal to expected:

<img width="1000" alt="NthReturn false 9 12 + 2" src="https://user-images.githubusercontent.com/11862657/61418511-bda2cf00-a8c8-11e9-8ac0-08291d5fece7.png">

thymikee commented:
Don't you think "Received value: " with empty space is a bit confusing at first?
Maybe:
```
Expected value: undefined
Received values
             1: "some"
			 2: null
```

or use different delimiter for numbers:

```
Expected value: undefined
Received value:
             1) "some"
			 2) null
```

For "not" matchers, could we dim noisy return values and only show the not-expected value in red? (similar for *NthReturned)

```
Expected value: not undefined
Received value:
             1: <dim>"some"</dim>
			 2: <red>undefined</red>
```

I'm also not sold on the "Call n", it's actually harder for me to get the error at first. How about:

```
Expected value: <green>89</green> for 9th call
Received value:
             8: <dim>"some"</dim>
			 9: <red>"rome"</red>
            10: <dim>89</dim>
```

pedrottimark commented:
Thank you for sharing your first impressions. By this time, they are long gone for me.

If there is only one call, and it is the last or n = 1, report can display received value on one line:

<img width="1000" alt="NthReturn false 1 1 b" src="https://user-images.githubusercontent.com/11862657/61491600-90642880-a97d-11e9-8869-a9b2d315039c.png">

For negative assertion, go one step farther to omit received value if it has same serialization:

<img width="1000" alt="NthReturn true 1 1 b" src="https://user-images.githubusercontent.com/11862657/61491758-e8029400-a97d-11e9-810e-0e61b75d8dcd.png">

What do you think about this alternative to original `Call n`?

<img width="1000" alt="NthReturn false 9 12 b" src="https://user-images.githubusercontent.com/11862657/61491821-0799bc80-a97e-11e9-8a2f-a447e5213f78.png">

And if additional values seem like noise instead of context, then I am happy to get rid of them!

thymikee commented:
Like it 👍

pedrottimark commented:
Updated pictures in preceding comments according to changes from https://github.com/facebook/jest/pull/8710#issuecomment-512704847 by Michał:

* if there is only one call for negative assertion, omit received value is it has same serialization as expected value
* otherwise, if there is only one call, display `Received value:` on one line
* otherwise display `Received value` without colon above one or more lines with values
* replace `Call n:` with `n:` label
* for last or nth matchers, display `->` at beginning of line with value of the expected call

Updated 94 snapshots

| | long name | short name |
| ---: | :--- | :--- |
| 13 | `toHaveReturnedWith` | `toReturnWith` |
| 13 | `toHaveLastReturnedWith` | `lastReturnedWith` |
| 21 | `toHaveNthReturnedWith` | `nthReturnedWith` |

jeysal commented on packages/expect/src/__tests__/__snapshots__/spyMatchers.test.js.snap:
```diff
@@ -127,44 +127,54 @@ Expected mock function to have been last called with:
-  <green>undefined</>
-But the last call <red>threw an error</>"
+Expected value: <green>undefined</>
+Received value: threw an error
```
Any chance we could make this "Received an error"?

pedrottimark replied:
Ouch, the special case takes a step forward and backward.

In a case not in tests nor pictures, what do you think about the phrases when a call is implied:

```sh
Expected value: undefined
Received value
             1: has not returned yet
             2: threw an error
```

pedrottimark replied:
Let’s see if we can improve the special strings independent of the labels.

It seems like the phrases need an explicit subject. I’m not sure if it is call or function.

Baseline was:
* `But the X call has not returned yet`
* `But the X call threw an error`

jeysal replied:
Oh I see, it needs to be compatible after both `Received:` and `1:`.
What about
```
Received an error

Received value 42

Received no return yet

Received
      1: an error
      2: value 42
      3: no return yet

```

Although we'd probably need a special case for the very common case where every call returned:
```
Received value
            1: 42
            2: 1337
```

jeysal replied:
Maybe I'm trying too hard to make everything a complete sentence when it's not even on the same line in the call list. Probably doesn't matter, the call list is super cool either way. @thymikee wdyt about the phrasing in this PR?

pedrottimark replied:
@jeysal Do you think the additional call lines seem relevant when test fails? The most helpful question from the best reviewer of my technical writing was “why are you telling this?”

Whatever we do here, we can use as pattern in next PR for `toHaveBeen*Called` matchers.

jeysal replied:
Yeah I think showing other calls around the call being asserted on and calls with equal values are good ideas

jeysal commented on packages/expect/src/__tests__/__snapshots__/spyMatchers.test.js.snap:
```diff
@@ -177,92 +187,86 @@ Received has value: <red>[Function fn]</>"
-  <green>\\"foo\\"</>
-But it last returned exactly:
-  <red>\\"foo\\"</>"
+Expected value: not <green>\\"foo\\"</>
```
These two lines are confusing, it sounds a bit like the number of returns being 1 is the problem. Not sure how to improve it though

pedrottimark replied:
Reading through your eyes always helps me. Unlike reports which contrast `Expected number of …` and `Received number of …` what do you think about a shorter label in these reports:

```sh
Number of returns: 1
```

jeysal replied:
Yes, I think that'd help!

pedrottimark commented:
Updated pictures in preceding comments according to changes after review by Tim:

* for these matchers, replace `Received number of returns` with `Number of returns`
* replace `Expected value` and `Received value` with `Expected` and `Received`
* replace `has not returned yet` with `function call has not returned yet`
* replace `threw an error` with `function call threw an error`
* in future PR, replace `called with no arguments` with `function called with no arguments`

The result is shorter labels in all reports and longer strings for special cases

Updated 102 snapshots

| | long name | short name |
| ---: | :--- | :--- |
| 15 | `toHaveReturnedWith` | `toReturnWith` |
| 15 | `toHaveLastReturnedWith` | `lastReturnedWith` |
| 21 | `toHaveNthReturnedWith` | `nthReturnedWith` |

pedrottimark commented:
Updated pictures in preceding comments according to changes after review by Tim:

* for these matchers, replace `Received number of returns` with `Number of returns`
* replace `Expected value` and `Received value` with `Expected` and `Received`
* replace `has not returned yet` with `function call has not returned yet`
* replace `threw an error` with `function call threw an error`
* in future PR, replace `called with no arguments` with `function called with no arguments`

The result is shorter labels in all reports and longer strings for special cases

Updated 102 snapshots

| | long name | short name |
| ---: | :--- | :--- |
| 15 | `toHaveReturnedWith` | `toReturnWith` |
| 15 | `toHaveLastReturnedWith` | `lastReturnedWith` |
| 21 | `toHaveNthReturnedWith` | `nthReturnedWith` |
